### PR TITLE
Don't share `JavaObjectSerializationCodec`

### LINKED ISSUE #26400: Don't share stateful codecs
We have some stateful codecs like `reentrant(Codec)`, these cannot be shared between projects during parallel sync because of concurrent state modification.

`JavaObjectSerializationCodec` is stateful and shared, what is wrong

### PR DESCRIPTION
**Pull request by 6hundreds**
Fixes #26400

### CODE REVIEW

**Comment by 6hundreds** [comment]
@bot-gradle test acc

File: subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/Codecs.kt [hunk 1]
```diff
@@ -232,17 +231,19 @@ class Codecs(
     }
 
     private
-    fun Bindings.completeWithReentrantBeanCodec() = append {
-        // This protects the BeanCodec against StackOverflowErrors but
+    fun Bindings.completeWithStatefulCodecs() = append {
+        bind(JavaObjectSerializationCodec(javaSerializationEncodingLookup))
```
Comment (bamboo) [inline subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/Codecs.kt:235]: 🤔 I'm wondering if it wouldn't be better for loading performance (more sharing of these reflection lookup results) to synchronize access to `readResolveMethod` and `readObjectHierarchy` instead.

I think we should benchmark one option against the other using Gradle profiler. WDYT?

Comment (6hundreds) [reply subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/Codecs.kt:235]: See https://github.com/gradle/gradle/issues/26400#issuecomment-1729275552

Comment (6hundreds) [reply subprojects/configuration-cache/src/main/kotlin/org/gradle/configurationcache/serialization/codecs/Codecs.kt:235]: @bamboo So we are ok to land it, right?

**Review by bamboo — APPROVED** [review]
Yes, let's revisit this choice once we can run more meaningful benchmarks.

**Comment by 6hundreds** [comment]
@bot-gradle test and merge

**Pull request merged**